File: Suho/과제진행하기/과제진행하기.py
def solution(plans):
    answer = []
    #선점스케줄링 / 스택 쓰는 문제
    #시간 -> 분으로 계산해서 시간 빼서 그 시간만큼 전작업 가능
    # 리스트의 [1]를 기준으로 sort해서 정렬하기.
    converted_plans =[]
    plans.sort(key=lambda x: x[1])
    #print(plans)
    for name, start, time in plans:
        #print(name,start,time)
        h, m = map(int,start.split(":"))
        start_m = h * 60 + m
        converted_plans.append([name,start_m,int(time)])
        
    pending_tasks =[] #잠시 멈춘 과제: [과제명, 남은 시간]
    
    #해당 과제의 time에서 rest_time을 스택에 같이 넣어두기
    for i in range(len(converted_plans) - 1):
        current_name, current_start, current_time = converted_plans[i]
        next_name, next_start, next_time = converted_plans[i + 1]

        
        # 다음 과제가 시작될 때까지 사용할 수 있는 시간
        available_time = next_start - current_start
        
        # 현재 과제를 끝내지 못함
        if current_time > available_time:
            
            #rest_time = 기존작업시각 - 다음에 온 과제 시각
            rest_time = current_time - available_time
            pending_tasks.append([current_name, rest_time])

    # 현재 과제를 완료함    
        else: 
            answer.append(current_name)

            # 과제를 끝내고 남은 시간
            extra_time = available_time - current_time

            # 남은 시간 동안 멈춰둔 과제 수행
            while pending_tasks and extra_time > 0:
                pending_name, pending_time = pending_tasks.pop()

                if pending_time <= extra_time:
                    extra_time -= pending_time
                    answer.append(pending_name)
                else:
                    pending_tasks.append([
                        pending_name,
                        pending_time - extra_time
                    ])
                    extra_time = 0

    #마지막 과제는 방해하는 다음 과제가 없으므로 완료
    answer.append(converted_plans[-1][0])

    #멈춘 과제를 최근에 멈춘 순서부터 완료
    while pending_tasks:
        name, remaining_time = pending_tasks.pop()
        answer.append(name)
    return answer

File: Jongwon/과제진행하기/과제진행하기.java
/*
과제는 시간이 되면 시작
새로운 과제 할 시간 -> 진행중인 과제가 있다면 멈춤
진행 중인 과제 끝내고 멈춘 과제 이어함
만약 과제를 끝낸 시간에 새로 시작해야 되는 과제와 잠시 멈춰둔 과제가 모두 있다면,
새로 시작해야 하는 과제부터 진행
그냥 새로 시작하는게 우선
과제 계획을 담은 이차원 문자열 배열 plans
과제를 끝낸 순서대로 이름을 배열에 담아 return
plans의 원소는 [name, start, playtime]의 구조
시간을 먼저 파악해서 제일 빠른거 먼저 실행
*/

import java.util.*;
class Solution {
    public String[] solution(String[][] plans) {
        String[] answer = {};
        int n = plans.length;

        Deque<String[]> stop = new ArrayDeque<>();
        List<String> task = new LinkedList<>();

        // 시간을 정렬함
        Arrays.sort(plans, (a,b) -> a[1].compareTo(b[1]));
        System.out.println(Arrays.deepToString(plans));

        // 현재 과제 종료시간이랑 다음 과제 시작시간을 비교
        for(int i=0; i<n-1; i++){
            // 시간을 :으로 분리 후 분으로 변환
            String nowTime = plans[i][1];
            String nextTime = plans[i+1][1];

            String[] now = nowTime.split(":");
            String[] next = nextTime.split(":");

            int nowHour = Integer.parseInt(now[0]);
            int nowMin = Integer.parseInt(now[1]);

            int nextHour = Integer.parseInt(next[0]);
            int nextMin = Integer.parseInt(next[1]);

            int nowSum = nowHour * 60 + nowMin;
            int nextSum = nextHour * 60 + nextMin;

            int playtime = Integer.parseInt(plans[i][2]);

            int finishTime = nowSum + playtime;
            int leftTime = nowSum + playtime - nextSum; // 현재 과제 남은 시간
            int remainTime = nextSum - finishTime; //다음 과제 시작 전까지 남는 시간

            // 현재 과제 종료시간 < 다음 과제 시작시간(현재 과제 끝냄)
            // 현재 과제 종료시간 > 다음 과제 시작시간(현재 과제 멈춤)
            // 현재 과제 스택에 넣기
            if(finishTime <= nextSum){
                task.add(plans[i][0]);

                while(remainTime > 0 && !stop.isEmpty()){
                    String[] stopPlans = stop.pop();

                    String name = stopPlans[0];
                    int time = Integer.parseInt(stopPlans[1]);

                    // 멈춤과제 끝냄
                    if(time <= remainTime){
                        task.add(name);
                        remainTime = remainTime-time;
                    }else{
                        stop.push(new String[]{
                                name,
                                String.valueOf(time - remainTime)
                        });
                        remainTime = 0;
                    }

                }

            }else{
                stop.push(new String[]{
                        plans[i][0],
                        String.valueOf(leftTime)
                });
            }
        }
        // 마지막 과제 처리
        task.add(plans[n - 1][0]);


        // 남아있는 멈춘 과제 처리
        while (!stop.isEmpty()) {
            task.add(stop.pop()[0]);
        }


        return task.toArray(new String[0]);
    }
}